engine/execution/messages.go

```diff
@@ -11,38 +11,118 @@ import (
+	return &ComputationResult{
+		ExecutableBlock:         block,
+		BlockExecutionResults:   NewEmptyBlockExecutionResults(block),
+		BlockAttestationResults: NewEmptyBlockAttestationResults(block),
+	}
+}
+
+func (cr *ComputationResult) CollectionExecutionResult(colIndex int) *CollectionExecutionResult {
+	if colIndex < 0 && colIndex > len(cr.StateSnapshots) {
+		return nil
+	}
+	return &CollectionExecutionResult{
+		blockHeader: cr.Block.Header,
+		collection: &flow.Collection{
+			Transactions: cr.CollectionAt(colIndex).Transactions,
+		},
+		collectionIndex:        colIndex,
+		isSystemCollection:     colIndex == len(cr.CompleteCollections),
+		updatedRegisters:       cr.StateSnapshots[colIndex].UpdatedRegisters(),
+		readRegisterIDs:        cr.StateSnapshots[colIndex].ReadRegisterIDs(),
+		emittedEvents:          cr.Events[colIndex],
+		transactionResults:     cr.transactionResultsByCollectionIndex(colIndex),
+		executionSnapshot:      cr.StateSnapshots[colIndex],
+		serviceEvents:          cr.ServiceEvents[colIndex],
+		convertedServiceEvents: cr.ConvertedServiceEvents[colIndex],
+	}
+}
+
+// BlockExecutionResults captures artifacts of execution of block collections
+type BlockExecutionResults struct {
 	StateSnapshots         []*state.ExecutionSnapshot
 	Events                 []flow.EventsList
-	EventsHashes           []flow.Identifier
-	ServiceEvents          flow.EventsList
+	ServiceEvents          []flow.EventsList
+	ConvertedServiceEvents []flow.ServiceEventList
 	TransactionResults     []flow.TransactionResult
 	TransactionResultIndex []int
 
 	// TODO(patrick): switch this to execution snapshot
 	ComputationIntensities meter.MeteredComputationIntensities
-
-	ChunkDataPacks []*flow.ChunkDataPack
-	EndState       flow.StateCommitment
-
-	*execution_data.BlockExecutionData
-	*flow.ExecutionReceipt
 }
 
-func NewEmptyComputationResult(
+func NewEmptyBlockExecutionResults(
 	block *entity.ExecutableBlock,
-) *ComputationResult {
+) *BlockExecutionResults {
 	numCollections := len(block.CompleteCollections) + 1
-	return &ComputationResult{
-		ExecutableBlock:        block,
+	return &BlockExecutionResults{
 		StateSnapshots:         make([]*state.ExecutionSnapshot, 0, numCollections),
 		Events:                 make([]flow.EventsList, numCollections),
-		EventsHashes:           make([]flow.Identifier, 0, numCollections),
-		ServiceEvents:          make(flow.EventsList, 0),
+		ServiceEvents:          make([]flow.EventsList, numCollections),
+		ConvertedServiceEvents: make([]flow.ServiceEventList, numCollections),
 		TransactionResults:     make([]flow.TransactionResult, 0),
 		TransactionResultIndex: make([]int, 0),
 		ComputationIntensities: make(meter.MeteredComputationIntensities),
-		ChunkDataPacks:         make([]*flow.ChunkDataPack, 0, numCollections),
-		EndState:               *block.StartState,
+	}
+}
+
+func (er *BlockExecutionResults) transactionResultsByCollectionIndex(colIndex int) []flow.TransactionResult {
```

Comment:
can we group transaction result by collection, so that we can get rid of or simplify this function?

Reply:
Thats what the next set of PRs are about.